lib/backend/k8sv2/k8s.go

```diff
@@ -0,0 +1,795 @@
+type KubeClient struct {
+	// Main Kubernetes clients.
+	clientSet *kubernetes.Clientset
+
+	// Client for interacting with CustomResourceDefinition.
+	crdClientV1 *rest.RESTClient
+
+	disableNodePoll bool
+
+	// Contains methods for converting Kubernetes resources to
+	// Calico resources.
+	converter Converter
+
+	// Clients for interacting with Calico resources.
+	bgpPeerClient       resources.K8sResourceClient
+	bgpConfigClient     resources.K8sResourceClient
+	nodeBgpConfigClient resources.K8sResourceClient
+	felixConfigClient   resources.K8sResourceClient
+	clusterInfoClient   resources.K8sResourceClient
+	ipPoolClient        resources.K8sResourceClient
+	gnpClient           resources.K8sResourceClient
+	nodeClient          resources.K8sResourceClient
+}
+
+func NewKubeClient(kc *apiconfig.KubeConfig) (api.Client, error) {
+	// Use the kubernetes client code to load the kubeconfig file and combine it with the overrides.
+	log.Debugf("Building client for config: %+v", kc)
+	configOverrides := &clientcmd.ConfigOverrides{}
+	var overridesMap = []struct {
+		variable *string
+		value    string
+	}{
+		{&configOverrides.ClusterInfo.Server, kc.K8sAPIEndpoint},
+		{&configOverrides.AuthInfo.ClientCertificate, kc.K8sCertFile},
+		{&configOverrides.AuthInfo.ClientKey, kc.K8sKeyFile},
+		{&configOverrides.ClusterInfo.CertificateAuthority, kc.K8sCAFile},
+		{&configOverrides.AuthInfo.Token, kc.K8sAPIToken},
+	}
+
+	// Set an explicit path to the kubeconfig if one
+	// was provided.
+	loadingRules := clientcmd.ClientConfigLoadingRules{}
+	if kc.Kubeconfig != "" {
+		loadingRules.ExplicitPath = kc.Kubeconfig
+	}
+
+	// Using the override map above, populate any non-empty values.
+	for _, override := range overridesMap {
+		if override.value != "" {
+			*override.variable = override.value
+		}
+	}
+	if kc.K8sInsecureSkipTLSVerify {
+		configOverrides.ClusterInfo.InsecureSkipTLSVerify = true
+	}
+	log.Debugf("Config overrides: %+v", configOverrides)
+
+	// A kubeconfig file was provided.  Use it to load a config, passing through
+	// any overrides.
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&loadingRules, configOverrides).ClientConfig()
+	if err != nil {
+		return nil, resources.K8sErrorToCalico(err, nil)
+	}
+
+	// Create the clientset
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, resources.K8sErrorToCalico(err, nil)
+	}
+	log.Debugf("Created k8s clientSet: %+v", cs)
+
+	crdClientV1, err := buildCRDClientV1(*config)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to build V1 CRD client: %s", err)
+	}
+
+	kubeClient := &KubeClient{
+		clientSet:       cs,
+		crdClientV1:     crdClientV1,
+		disableNodePoll: kc.K8sDisableNodePoll,
+	}
+
+	// Create the Calico sub-clients.
+	kubeClient.ipPoolClient = resources.NewIPPoolClient(cs, crdClientV1)
+	kubeClient.nodeClient = resources.NewNodeClient(cs, crdClientV1)
+	kubeClient.gnpClient = resources.NewGlobalNetworkPolicyClient(cs, crdClientV1)
```

Comment:
I think we need one for NetworkPolicy too, and then combine the Calico CRD-backed NetworkPolicy resources with the k8s Network Policy.  Arguably this could fall out into a different task though - as long as we are tracking that.

Reply:
Yup. I realized that I missed out the network policy conversion bit. I'll push this with some shortly.

Reply:
So, a couple of things aren't clear to me w.r.t policies in the backend.

Is the CRD-backed NetworkPolicy resource the GlobalNetworkPolicy or is it a Calico Policy (just to distinguish from K8s NetworkPolicy) but is allowed to be namespaced?
My understanding was that we have two v2 policy related resources: `GlobalNetworkPolicy` and `NetworkPolicy` and that the `GlobalNetworkPolicy` resource will be backed by a CRD and `NetworkPolicy` is backed by K8s _NetworkPolicy_. 

In the current (corresponding to the v1 style backend) _GET_/_LIST_ for a _Policy_ resource, the k8s backend combines `GlobalNetworkPolicy` (a CRD) and `NetworkPolicy` (K8s API). With the v2 datamodel, with 2 separate client resources/APIs, does this combining still make sense?

Reply:
So for v3.0 the network policy that is sync'd to Felix will come from 3 places in KDD:
-  K8s Network Policy
-  Calico Network Policy (CRD)
-  Calico Global Network Policy (CRD)

The names/selectors are adjusted on the fly in the Felix syncer, such that Felix itself isn't actually aware of these different sources of policy.

Regarding CRUD:
-  GlobalNetworkPolicy maps straight through to the CRD
-  NetworkPolicy maps through to both the Kubernetes NP and the Calico NP CRD  - so it's the union of the two - with a name prefix ensuring name uniqueness across the two sources.

Reply:
Got it.

Reply:
Fixing up the syncer is a separate task - it might prove to be easier to use the new syncer infrastructure that i've implemented -- but would require us to implement watcher support in kdd.